File: delete-node-in-a-bst.py
# Runtime: 103 ms
class Solution(object):

    def findMin(self, root):
        while root.left:
            root = root.left
        return root

    def deleteNode(self, root, key):
        if not root:
            return root
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)
        elif root.val > key:
            root.left = self.deleteNode(root.left, key)
        else:
            if not root.left and not root.right:
                root = None
            elif not root.left:
                root = root.right
            elif not root.right:
                root = root.left
            else:
                temp = self.findMin(root.right)
                root.val = temp.val
                root.right = self.deleteNode(root.right, temp.val)

        return root

File: delete-node-in-a-bst.scala
//Runtime: 2156 ms
object Solution {

  def findMin(root: TreeNode): Int = {
    if (root.left == null) root.value
    else findMin(root.left)
  }

  def deleteNode(root: TreeNode, key: Int): TreeNode = {
    if (root == null) root
    else if (root.value > key) root.left = deleteNode(root.left, key)
    else if (root.value < key) root.right = deleteNode(root.right, key)
    else {
      if (root.left == null && root.right == null) return null
      else if (root.left == null && root.right != null) return root.right
      else if (root.right == null && root.left != null) return root.left
      else {
        val min = findMin(root.right)
        root.value = min
        root.right = deleteNode(root.right, min)
      }
    }
    root
  }
}
